ooni/tests/test_scheduler.py

```diff
@@ -274,43 +274,97 @@ def test_deck_run_twice(self, mock_resources, mock_probe_ip,
+                        if last_run == '2000-01-02T08:00:00Z':
+                            zero += 1
+            self.assertGreater(hourly, 0)
+            self.assertGreater(daily, 0)
+            self.assertLess(zero, hourly)
+            self.assertLessEqual(zero, 1) # should ALMOST never happen
+
+            # leaping to the end of the day
+            dummy_clock.advance((datetime(2000,01,02,23,59,59) - FakeDatetime.utcnow()).total_seconds())
+            for t in scheduler_service._scheduled_tasks:
+                with open(os.path.join(self.scheduler_directory, t.identifier)) as in_file:
+                    if t.schedule == '@daily':
+                        self.assertEqual(in_file.read(), '2000-01-02T07:00:45Z')
+                    elif t.schedule == '@hourly':
+                        self.assertEqual(in_file.read(), '2000-01-02T23:59:59Z')
+
+            # save ~30% of the testcase runtime while ticking through six hours
+            for t in scheduler_service._scheduled_tasks[:]:
+                if t.schedule == '@hourly':
+                    scheduler_service.unschedule(t)
+
+            # ticking through six hours
+            dummy_clock.pump([random.uniform(0, 120) for i in xrange(6*60)])
+            for t in scheduler_service._scheduled_tasks:
+                with open(os.path.join(self.scheduler_directory, t.identifier)) as in_file:
+                    # randomized clock kills 30s resolution of the scheduler
+                    self.assertRegexpMatches(in_file.read(), '^2000-01-03T0[012]:..:..Z$')
+            self.assertGreater(FakeDatetime.utcnow(), datetime(2000,1,3, 5,0,0)) # should be ~6:00
+
+            # verify, that double-run does not happen even in case of reseeding (reboot/restart)
+            dummy_clock.advance((datetime(2000,01,03,23,59,59) - FakeDatetime.utcnow()).total_seconds())
+            launches = {}
+            while FakeDatetime.utcnow() < datetime(2000,1,4, 6,0,0):
+                dummy_clock.advance(random.uniform(0, 120))
+                for t in scheduler_service._scheduled_tasks:
+                    with open(os.path.join(self.scheduler_directory, t.identifier)) as in_file:
+                        launches.setdefault(t.identifier, set())
+                        launches[t.identifier].add(in_file.read())
+                    self.assertLessEqual(t._smear_coef, 1.0)
+                    t._smear_coef = random.random()
+            self.assertEqual(len(launches), len(scheduler_service._scheduled_tasks))
+            self.assertEqual({k: len(v) for k, v in launches.iteritems()}, dict.fromkeys(launches.iterkeys(), 2))
 
             # We check that the run method of the deck was called twice
+            # NB: That does NOT check that @daily task was called exactly twice
             self.mock_deck.run.assert_has_calls([
                 mock.call(mock_director), mock.call(mock_director)
             ])
```

Comment:
yes you are right. This is just asserting that is was "called **at least** twice", but not exactly twice.

We can enforce exactly twice by doing below:

```
self.assertEqual(self.mock_deck.run.call_count, 2)
```